services/httpd/handler.go

```diff
@@ -993,8 +1012,366 @@ func fixLiterals(node influxql.Node) {
+			return
+		}
+	}
+
+	if bucket := h.oneBucket(w, fmt.Sprintf("%s/%s", db, rp)); bucket != nil {
+		b, err := json.Marshal(bucket)
+		if err != nil {
+			h.httpError(w, fmt.Sprintf("retrieve bucket marshaling error: %s", err.Error()), http.StatusBadRequest)
+			return
+		}
+		if _, err := w.Write(b); err != nil {
+			h.httpError(w, fmt.Sprintf("retrieve bucket error writing response: %s", err.Error()), http.StatusBadRequest)
+			return
+		}
+	}
+	return
+}
+
+func (h *Handler) serveBucketListV2(w http.ResponseWriter, r *http.Request, user meta.User) {
+	var err error
+	var db, rp, after string
+	limit := 20
+	offset := 0
+
+	if h.Config.AuthEnabled {
+		if user == nil {
+			h.httpError(w, "list buckets - user is required", http.StatusForbidden)
+			return
+		}
+	}
+
+	// This is the API for returning an array of buckets
+	id := r.URL.Query().Get("id")
+	name := r.URL.Query().Get("name")
+	if id != "" || name != "" {
+		if id != "" && name != "" && id != name {
+			h.httpError(w, fmt.Sprintf("list buckets: name: %q and id: %q do not match", name, id), http.StatusBadRequest)
+			return
+		} else if id != "" {
+			name = id
+		}
+		if b := h.oneBucket(w, name); b != nil {
+			sendBuckets(w, []Bucket{*b})
+		}
+		return
+	}
+
+	if after = r.URL.Query().Get("after"); after != "" {
+		db, rp, err = bucket2dbrp(after)
+		if err != nil {
+			h.httpError(w, fmt.Sprintf("list buckets invalid parameter - after=%q: %s", after, err.Error()), http.StatusNotFound)
+			return
+		}
+	}
+
+	if limitStr := r.URL.Query().Get("limit"); limitStr != "" {
+		if l, err := strconv.ParseInt(limitStr, 10, 64); err != nil {
+			h.httpError(w, fmt.Sprintf("list buckets parameter is not an integer - limit=%q: %s", limitStr, err.Error()), http.StatusBadRequest)
+			return
+		} else {
+			limit = int(l)
+		}
+	}
+
+	if offsetStr := r.URL.Query().Get("offset"); offsetStr != "" {
+		if after != "" {
+			h.httpError(w, "list buckets cannot have both \"offset\" and \"after\" arguments", http.StatusBadRequest)
+			return
+		} else if o, err := strconv.ParseInt(offsetStr, 10, 64); err != nil {
+			h.httpError(w, fmt.Sprintf("list buckets parameter is not an integer - offset=%q: %s", offsetStr, err.Error()), http.StatusBadRequest)
+			return
+		} else {
+			offset = int(o)
+		}
+	}
+
+	dbIndex := 0
+	rpIndex := 0
+	dbs := h.MetaClient.Databases()
+	if after != "" {
+		if dbIndex, rpIndex = findBucketIndex(dbs, db, rp); dbIndex < 0 {
+			h.httpError(w, fmt.Sprintf("list buckets \"after\" parameter not found: %q", after), http.StatusNotFound)
+			return
+		}
+	} else if offset > 0 {
+		if dbIndex, rpIndex = findBucketOffsetIndex(dbs, offset); dbIndex < 0 {
+			// TODO (DSB): Is this the right error?
+			http.Error(w, fmt.Sprintf("list buckets offset past end of list: %d", offset), http.StatusNotFound)
+			return
+		}
+	}
+
+	buckets := make([]Bucket, 0, limit)
+
+outer:
+	for ; dbIndex < len(dbs); dbIndex++ {
```

Comment:
We should sync with the UI team on pagination of these API's - the code currently assumes you might have too many buckets to fit in memory and so need server-side pagination, but I don't think that's really a use case we support. In general I think the UI has been moving from server-side pagination to client-side pagination, for simplicity.

Reply:
Asked in ui slack channel. Let's proceed with the currently documented behavior and modify if they get back to us.